core/src/main/scala/org/apache/spark/sql/TiExtensions.scala

```diff
@@ -34,14 +38,20 @@ class TiExtensions extends (SparkSessionExtensions => Unit) {
 
   // call from pyspark only
   def getOrCreateTiContext(sparkSession: SparkSession): TiContext =
-    synchronized {
-      tiContextMap.get(sparkSession) match {
-        case Some(tiContext) => tiContext
-        case None =>
-          // TODO: make Meta and RegionManager independent to sparkSession
-          val tiContext = new TiContext(sparkSession)
-          tiContextMap.put(sparkSession, tiContext)
-          tiContext
+    // check catalog plugin mode
+    if (!TiExtensions.validateCatalog(sparkSession)) {
```

Comment:
In my opinion, catalog validation should be added to apply(), since getOrCreateTiContext invoke would happen after Rule construction.

Reply:
It may not suitable if add catalog validation to the existing apply(), so, i did that before. Maybe we can add a inject method before others.